### PR TITLE
Add option to specify port-range for outgoing NAT.

### PR DESCRIPTION
**Pull request by databus23**
This commit allow to optionally specify a port range for the outgoing NAT rules that are created by felix.

## Description
This option allows to specify the port range used by iptables for outgoing NAT (e.g. `MASQUERADE`). We are running calico on a VMware based OpenStack with security groups implemented by a stateless firewall (VMware distributed virtual switch, DVS). This imposes some limitations how `egress` rules are implemented. In fact we need to ensure that outgoing traffic is coming from the default ephemeral port range (32768-65535). This works fine for traffic initiated by processes from the host network but breaks for the default masquerading case. The reason being that iptables picks arbitrary ports to resolve port collisions by default.

I realise that this is a rather exotic issue but the proposed configuration change is rather generic and fully optional.
I'm hoping to get this merged so that we can use upstream calico in our stack.  Would this change be generally acceptable to you?

## Todos
- [x] Unit tests (full coverage)
- [x] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by fasaxc** [comment]
Sounds like a reasonable tweak to me, to accept it, I think we'd need these other changes:

- Add UTs in the rules package to check that the new rules render as expected.
- So this can be configured with calicoctl, it needs to be exposed in libcalico-go here: https://github.com/projectcalico/libcalico-go/blob/master/lib/apis/v3/felixconfig.go
- So other people can find it, it needs to be added to the docs here: https://github.com/projectcalico/calico/blob/master/master/reference/felix/configuration.md

**Comment by databus23** [comment]
@fasaxc Great! I added a unit test, verified it manually and added to more PR as requested.

File: rules/nat_test.go [hunk 1]
```diff
@@ -55,6 +56,37 @@ var _ = Describe("NAT", func() {
 			},
 		}))
 	})
+	It("should render rules when active with explicit port range", func() {
+
+		//copy struct
+		localConfig := rrConfigNormal
+		localConfig.NATPortRange = numorstring.Port{99, 100, ""}
```
Comment (fasaxc) [inline rules/nat_test.go]: I think there's a utility function for making each type of port.  `numorstring.PortRange(a, b)`, maybe?

File: rules/nat.go [hunk 2]
```diff
@@ -26,13 +27,39 @@ func (r *DefaultRuleRenderer) NATOutgoingChain(natOutgoingActive bool, ipVersion
...
+						Protocol("tcp"),
+				},
+				{
+					Action: iptables.MasqAction{ToPorts: toPorts},
+					Match: iptables.Match().
+						SourceIPSet(masqIPsSetName).
+						NotDestIPSet(allIPsSetName).
+						Protocol("udp"),
+				},
+				{
```
Comment (fasaxc) [inline rules/nat.go]: Just occurred to me, is MASQUERADE a "final" action; once you've hit the tcp rule above, will the packet also hit this rule and get double NATted, or something?

**Comment by fasaxc** [comment]
libcalico-go PR: https://github.com/projectcalico/libcalico-go/pull/893

File: rules/nat.go (hunk 2 above)
Comment (databus23) [reply rules/nat.go]: Hmm, good question. I obviously thought it was. Let me verify that...

Comment (databus23) [reply rules/nat.go]: Of course its not. Good catch! I updated the code to exit early for tcp and udp.

**Comment by databus23** [comment]
@fasaxc I addressed all your feedback in all three PRs.

File: config/config_params.go [hunk 3]
```diff
@@ -180,6 +180,7 @@ type Config struct {
 	FailsafeOutboundHostPorts []ProtoPort `config:"port-list;udp:53,udp:67,tcp:179,tcp:2379,tcp:2380,tcp:6666,tcp:6667;die-on-fail"`
 
 	KubeNodePortRanges []numorstring.Port `config:"portrange-list;30000:32767"`
+	NatPortRange       numorstring.Port   `config:"portrange;;local"`
```
Comment (fasaxc) [inline config/config_params.go]: Just spotted the `local` on here; that shouldn't be needed - no reason to limit this config to env vars only

Comment (fasaxc) [reply config/config_params.go]: Maybe you added this to get the tests to pass without the libcalico-go change?

Comment (databus23) [reply config/config_params.go]: Yes, I added that to make the tests pass. I removed it as requests.

**Comment by databus23** [comment]
@fasaxc all done

**Pull request merged**